posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -501,27 +504,45 @@ Calculation time: 1.345
 
 ----
 
+Note that we set the number of _chunks_ to 64 in this run.
+
 What we can observe is that, with 3,008 terms, we have correctly calculated more
 than 3,600 digits accurately.
 
-If we set the size of the ForkJoinPool to 1, 2, 4, 8, 12, 16, 32 we get the
-following values for the calculation time (in seconds) respectively: 12.12,
-6.23, 3.03, 1.78, 1.34, 1.36, 1.43
+We can measure the effect of changing the Thread pool size on the total calculation time. The following graphs shows the relation between the two and it also shown another (orange) curve that plots the case when the system scales perfectly and unbounded.
 
-So, we see a near linear speed-up by going from one thread to two threads.
+image::2020-03-10-calculating-pi-concurrently/CalculationTimeTotalVersusThreadPoolSize.png[Measured versus theoretical scaling in function of Thread Pool size][{half-width}]
+
+So, we see a nearly linear speed-up by increasing the number of threads from 1 to 12.
 A further increase of the thread-count doesn't yield a further linear speed-up:
 this may be caused by different factors, not in the least by the fact that we
 have a single chip processor with a shared on-chip cache.
 Of course, since it's a six core CPU (with hyper-threads that don't yield the
 same performance as the regular CPU threads), we don't get a speed-up beyond 12
 threads in the ForkJoinPool.
 
+We can do another experiment by setting the number of _chunks_ and the number of threads
+in the thread pool to 12. One would expect that this should be no impact. In practice
+however, the total execution time increases from 1.37s to 3.31s! How can that be?
+
+The explanation can be found by looking at the following graph which plots the time to calculate each term in the series (and this for a run with 10,000 calculated terms).
+
+image::2020-03-10-calculating-pi-concurrently/CalculationTimePerTerm.png[Calculation time per term][{half-width}]
+
+We see that the time to calculate terms decreases significantly between the first term and term 4,300 at which point it reaches kind of a plateau.
+
+As we divide the calculation into chunks, the first few chunks will determine the overall speed at which the calculation is performed.
+
+It turns out that the choice of chunking the calculation is not optimal. Another approach is to spread the calculation across different threads in a round robin fashion as illustrated in an alternative implementation in the source code.
```

Comment:
So not every chunk takes as long and some threads will be idle?

Reply:
Akka Streams should lead to this 'round robin' behavior automatically, right?

Reply:
I've added more details on this + a graph that clearly illustrates what's causes the apparent slowdown when setting # chunks equal to # cores.